src/Microsoft.Win32.SystemEvents/src/Microsoft/Win32/SystemEvents.cs

```diff
@@ -0,0 +1,1497 @@
+            {
+                if (s_handlers == null)
+                {
+                    s_handlers = new Dictionary<object, List<SystemEventInvokeInfo>>();
+                    EnsureSystemEvents(false, false);
+                }
+
+                List<SystemEventInvokeInfo> invokeItems;
+
+                if (!s_handlers.TryGetValue(key, out invokeItems))
+                {
+                    invokeItems = new List<SystemEventInvokeInfo>();
+                    s_handlers[key] = invokeItems;
+                }
+                else
+                {
+                    invokeItems = s_handlers[key];
+                }
+
+                invokeItems.Add(new SystemEventInvokeInfo(value));
+            }
+        }
+
+        /// <devdoc>
+        ///      Console handler we add in case we are a console application or a service.
+        ///      Without this we will not get end session events.
+        /// </devdoc>
+        private bool ConsoleHandlerProc(int signalType)
+        {
+            switch (signalType)
+            {
+                case Interop.User32.CTRL_LOGOFF_EVENT:
+                    OnSessionEnded((IntPtr)1, (IntPtr)Interop.User32.ENDSESSION_LOGOFF);
+                    break;
+
+                case Interop.User32.CTRL_SHUTDOWN_EVENT:
+                    OnSessionEnded((IntPtr)1, (IntPtr)0);
+                    break;
+            }
+
+            return false;
+        }
+
+        private Interop.User32.WNDCLASS WndClass
+        {
+            get
+            {
+                if (s_staticwndclass == null)
+                {
+                    const string classNameFormat = ".NET-BroadcastEventWindow.{0}.{1}.{2}";
+
+                    IntPtr hInstance = Interop.Kernel32.GetModuleHandle(null);
+
+                    s_className = string.Format(System.Globalization.CultureInfo.InvariantCulture,
+                        classNameFormat,
+                        "4.0.0.0", //todo
+                        Convert.ToString(AppDomain.CurrentDomain.GetHashCode(), 16),
+                        s_domainQualifier);
+
+                    Interop.User32.WNDCLASS tempwndclass = new Interop.User32.WNDCLASS();
+                    tempwndclass.hbrBackground = (IntPtr)(Interop.User32.COLOR_WINDOW + 1);
+                    tempwndclass.style = 0;
+
+                    _windowProc = new Interop.User32.WndProc(this.WindowProc);
+                    tempwndclass.lpszClassName = s_className;
+                    tempwndclass.lpfnWndProc = _windowProc;
+                    tempwndclass.hInstance = hInstance;
+                    s_staticwndclass = tempwndclass;
+                }
+                return s_staticwndclass;
+            }
+        }
+
+        private IntPtr DefWndProc
+        {
+            get
+            {
+                if (s_defWindowProc == IntPtr.Zero)
+                {
+                    string defproc = (Marshal.SystemDefaultCharSize == 1 ? "DefWindowProcA" : "DefWindowProcW");
```

Comment:
I assume we don't need DefWindowProcA !

Reply:
So you can get rid of GetProcAddress and use real Pinvoke?

Reply:
This is not for PInvoke, this is to get a function pointer to pass to Set*Pointer.  See https://github.com/dotnet/corefx/pull/27420/files/d47f746995a990dc45cd5b1a2cc978a7e4d0ed5f#diff-e3aaa30d0638c60ada505f9818b480f7R561.

Reply:
Ah. We can remove DefWindowProcA though. The 90's want their export back.

Reply:
> The 90's want their export back.

This made me smile.